sklearn/experimental/_column_transformer.py

```diff
@@ -0,0 +1,427 @@
+
+    Read more in the :ref:`User Guide <column_transformer>`.
+
+    Parameters
+    ----------
+    transformers : list of tuples
+        List of (name, transformer, column) tuples specifying the transformer
+        objects to be applied to subsets of the data.
+        Specifying the column as a scalar will pass a 1d array or pandas Series
+        to the transformer. Specifying the columns as a list, int array, slice
+        or boolean mask will pass a 2d array or DataFrame. Strings can
+        reference columns if the input is a DataFrame, integers are always
+        interpreted as the positional columns.
+        When passing a single column to a transformer that expects 2D input
+        data, the column should be specified a list of one element.
+
+    n_jobs : int, optional
+        Number of jobs to run in parallel (default 1).
+
+    transformer_weights : dict, optional
+        Multiplicative weights for features per transformer.
+        Keys are transformer names, values the weights.
+
+    Attributes
+    ----------
+    transformers_ : list
+        The collection of fitted transformers as tuples of
+        (name, fitted_transformer, column)
+
+    Examples
+    --------
+    >>> from sklearn.experimental import ColumnTransformer
+    >>> from sklearn.preprocessing import Normalizer
+    >>> union = ColumnTransformer(
+    ...     [("norm1", Normalizer(norm='l1'), [0, 1]),
+    ...      ("norm2", Normalizer(norm='l1'), slice(2, 4))])
+    >>> X = np.array([[0., 1., 2., 2.],
+    ...               [1., 1., 0., 1.]])
+    >>> union.fit_transform(X)    # doctest: +NORMALIZE_WHITESPACE
+    array([[ 0. ,  1. ,  0.5,  0.5],
+           [ 0.5,  0.5,  0. ,  1. ]])
+
+    """
+
+    def __init__(self, transformers, n_jobs=1, transformer_weights=None):
+        self.transformers = tosequence(transformers)
+        self.n_jobs = n_jobs
+        self.transformer_weights = transformer_weights
+        self._validate_transformers()
+
+    @property
+    def _transformers(self):
+        """
+        Internal list of transformer only containing the name and
+        transformers, since BaseComposition expects lists of tuples of len 2
+        """
+        return [(name, trans) for name, trans, col in self.transformers]
+
+    @_transformers.setter
+    def _transformers(self, value):
+        self.transformers = [
+            (name, trans, col) for ((name, trans), (_, _, col))
+            in zip(value, self.transformers)]
+
+    def get_params(self, deep=True):
+        """Get parameters for this estimator.
+
+        Parameters
+        ----------
+        deep : boolean, optional
+            If True, will return the parameters for this estimator and
+            contained subobjects that are estimators.
+
+        Returns
+        -------
+        params : mapping of string to any
+            Parameter names mapped to their values.
+        """
+        return self._get_params('_transformers', deep=deep)
```

Comment:
we could alternatively have ``_get_params`` optionally take a list, and then we don't need to add this unintuitive property.
like
```python
def _get_params(self, attr, estimators=None, deep=True):
    if estimators is None:
        estimators = getattr(self, attr)
```

Reply:
I wonder, what is the importance of `# Ensure strict ordering of parameter setting`? Is it for cases when `set_params` sets sub-estimators and their parameters *in one go*? It looks like this clever trick of replacing `transformers` with `_transformers` breaks this behaviour.

For example, `ct.set_params(transformers='foo')` is supposed to be taken care of in part 1: https://github.com/scikit-learn/scikit-learn/blob/62a017efa047e9581ae7df8bbaa62cf4c0544ee4/sklearn/utils/metaestimators.py#L53-L71
but will actually be dealt with in part 3. This is because `attr=='_transformers'` is not a key in `params=={'transformers':'foo'}`.

In contrast, `pipe.set_params(steps='foo')` will be dealt with in part 1, because `attr=='steps'` is a key in `params=={'steps':'foo'}`.

Just saying. I doubt anyone found this to be a problem during these 4.5 years that passed.